### PR TITLE
Add memory format support to `empty_like` operator

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28544 Add torch.memory_format support to the TorchScript
* #28292 Add memory format support to the `resize_` op.
* #28291 TensroIterator preserve format for binary, ternary operators.
* #28077 Add channels last support to cuda.comm.scatter and gather
* #28076 Kill `operator==` of TensorOptions as confusing one
* #27979 Add memory format support to `resize_as_` operator
* #27891 Cleanup testing of _like operators
* #27890 Add memory format support to `randn_like` operator
* #27889 Add memory format support to `randint_like` operator
* #27562 Add memory format support to `zeros_like` operator
* #27561 Add memory format support to `rand_like` operator
* #27270 Add memory format support to `ones_like` operator
* #27262 Add memory format support to `full_like` operator
* **#27244 Add memory format support to `empty_like` operator**

Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980310](https://our.internmc.facebook.com/intern/diff/D17980310)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@houseroad I had to whitelist `empty_like` inside of check_backward_compatibility.py as it doen't like changing of default value 
```
processing existing schema:  aten::empty_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=0) -> (Tensor)
Can NOT find backward compatible schemas after changes for schema aten::empty_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=0) -> (Tensor) from the following candidates:
[
aten::empty_like(Tensor self, *, int? memory_format=None) -> (Tensor)
	aten::empty_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor)
]
```

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -1115,12 +1115,11 @@
...
   device_guard: False
 
-- func: empty_like(Tensor self) -> Tensor
-  use_c10_dispatcher: full
+- func: empty_like(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor
   device_guard: False
   supports_named_tensor: True
 
-- func: empty_like.dtype(Tensor self, *, ScalarType dtype, Layout layout, Device device, bool pin_memory=False, MemoryFormat? memory_format=contiguous_format) -> Tensor
+- func: empty_like.dtype(Tensor self, *, ScalarType dtype, Layout layout, Device device, bool pin_memory=False, MemoryFormat? memory_format=None) -> Tensor
```
Comment (houseroad) [inline aten/src/ATen/native/native_functions.yaml]: None and contigous_format (i.e., 0) have the same behavior?

Comment (VitalyFedyunin) [reply aten/src/ATen/native/native_functions.yaml]: Yes, but it is exceptional case. No need to handle it in the test.

Comment (houseroad) [reply aten/src/ATen/native/native_functions.yaml]: Sounds fine.

**Comment by VitalyFedyunin** [comment]
@ailzhang hi

File: aten/src/ATen/native/TensorFactories.cpp [hunk 2]
```diff
@@ -196,22 +198,19 @@ Tensor empty_like(
...
-    } else if (self.is_contiguous(MemoryFormat::Contiguous)) {
-      use_memory_format = MemoryFormat::Contiguous;
-    } else {
-      TORCH_CHECK(
-          false,
-          "undefined behavior of the preserve format, source tensor neither channels last nor contiguous")
+  if (self.is_quantized()) {
+
+    auto memory_format =
+        optional_memory_format.value_or(MemoryFormat::Contiguous);
```
Comment (ezyang) [inline aten/src/ATen/native/TensorFactories.cpp]: To be changed later right :>

File: tools/autograd/gen_variable_factories.py [hunk 3]
```diff
@@ -64,7 +64,10 @@ def process_function(decl, has_tensor_options, disable_autograd):
     requires_grad = "options.requires_grad()" if has_tensor_options else "false"
     if decl['name'].endswith('_like') and not has_tensor_options:
         # it's a tensor
-        actuals.append('{}.options().is_variable(false)'.format(actuals[0]))
+        if decl['name'] == 'empty_like':
```
Comment (ezyang) [inline tools/autograd/gen_variable_factories.py]: Umm... this looks very non-general. (I assume it gets fixed later in the stack but it is very shocking here)

File: test/test_torch.py [hunk 4]
```diff
@@ -12371,7 +12371,7 @@ def check_sum_all(tensor):
         else:
             check_sum_all(torch.tensor([True, False, True], dtype=torch.bool, device=device))
 
-    def _test_memory_format_transformations(self, device, input_generator_fn, transformation_fn):
+    def _test_memory_format_transformations(self, device, input_generator_fn, transformation_fn, compare_data=True):
```
Comment (ezyang) [inline test/test_torch.py]: Goofy but ok.

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/TensorFactories.cpp (hunk 2 above)
Comment (VitalyFedyunin) [reply aten/src/ATen/native/TensorFactories.cpp]: Yes, right now we need to introduce this option, after we check all current code for expected behaviour and change default.

File: tools/autograd/gen_variable_factories.py (hunk 3 above)
Comment (VitalyFedyunin) [reply tools/autograd/gen_variable_factories.py]: Yes, need to list by function until I convert all of them

File: aten/src/ATen/native/TensorFactories.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/native/TensorFactories.cpp]: If it's to be changed later, can we add a comment linking some issue or with some information ?

Comment (ezyang) [reply aten/src/ATen/native/TensorFactories.cpp]: What Richard said

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -1116,12 +1116,11 @@
 - func: empty.out(int[] size, *, MemoryFormat? memory_format=None, Tensor(a!) out) -> Tensor(a!)
   device_guard: False
 
-- func: empty_like(Tensor self) -> Tensor
-  use_c10_dispatcher: full
+- func: empty_like(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml]: With the benefit of hindsight, I should have asked: why do we have two overloads of `empty_like` in the first place?!

Comment (VitalyFedyunin) [reply aten/src/ATen/native/native_functions.yaml]: Do you mean this line and line 1123?

Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: yes. But see my most recent PR.

Comment (VitalyFedyunin) [reply aten/src/ATen/native/native_functions.yaml]: As far as I remember combining them together was never considered by me.

**Pull request closed without merging**